### ISSUE TITLE
Provide support for ARM64/M1 devices (Windows and macOS)

### ISSUE THREAD

**Issue by brunoborges** [LWJGL/lwjgl3#601]
Hi all,

I work on the Microsoft Java Engineering Group, and I'd like to openly discuss how we can help port LWJGL to Windows on ARM64 devices (e.g. Surface X) as well as macOS on Apple Silicon M1.

Best,
Bruno Borges

**Comment by Spasi** [comment]
Hey Bruno,

Porting LWJGL means getting the CI pipelines to build its native dependencies for the new platform/architecture. LWJGL has a simple core that's easy to build, but the different bindings depend on native libraries with varying implementation quality and build complexity. Some will be trivial to build, others may require simple fixes we can apply on our own, sometimes we may have to wait for the upstream maintainers to do the work. A few might even be impossible to build. In any case, we don't need every dependency to release LWJGL and our [build customizer](https://www.lwjgl.org/customize) can handle platforms/architectures with missing bindings.

I'm guessing that Microsoft is mostly interested in getting Minecraft to run on ARM64. We could focus on bindings that Minecraft uses first, deal with the rest later. Those most likely include: `GLFW`, `OpenGL`, `OpenAL` and `stb`. Could you please confirm and check if they use anything else?

We have a separate account for CI ([LWJGL-CI](https://github.com/LWJGL-CI?tab=repositories)). Each repository has a single "LWJGL CI" commit on top of the upstream commits. This commit includes the build scripts and any fixes that need to be applied. When the binding is updated, the CI commit is rebased on top of the new upstream commit. This keeps the history clean and enables interested users to easily identify in which state each binding has been built (it's always the `HEAD~1` commit). There are two interesting files in each CI commit: `.travis.yml` and `appveyor.yml` (note: we completely overwrite those files if they exist in the upstream project). We use Travis CI for Linux and macOS builds, AppVeyor for Windows builds. Each build file contains a matrix of builds for the various architectures. Artifacts are uploaded to LWJGL's S3 build bucket. Before opening a PR and while testing a new build, you probably want to comment out the `awscli` installation and `aws s3 cp` commands, to avoid build errors. Ideally, replace them with your own upload procedure to test the artifacts.

As a test, I got GLFW to build on Windows ARM64 and macOS ARM64. It was straightforward:

- [.travis.yml](https://github.com/LWJGL-CI/glfw/blob/c5ee7da34cad750429b1f2bea4c6ea7ed4e7de6f/.travis.yml#L305), build output [here](https://travis-ci.com/github/LWJGL-CI/glfw/jobs/445152654)
- [appveyor.yml](https://github.com/LWJGL-CI/glfw/blob/c5ee7da34cad750429b1f2bea4c6ea7ed4e7de6f/appveyor.yml#L19), build output [here](https://ci.appveyor.com/project/LWJGL-CI/glfw/builds/36431044/job/lq7uvhdqp9uuj8f9)

One problem is that I don't have any hardware to test on, but I'm getting:

`dumpbin /HEADERS glfw.dll | findstr machine`
> AA64 machine (ARM64)

for https://build.lwjgl.org/nightly/windows/arm64/glfw.dll and

`file libglfw.dylib`
> libglfw.dylib: Mach-O 64-bit dynamically linked shared library arm64

for https://build.lwjgl.org/nightly/macosx/arm64/libglfw.dylib.

**Comment by Spasi** [comment]
I've hit a showstopper. Looks like [dyncall](https://dyncall.org/) does [not support](https://ci.appveyor.com/project/LWJGL-CI/dyncall/build/job/vc9yq6m830xk5oby) Windows ARM64 yet. LWJGL uses dyncall to implement upcalls/callbacks from C to Java. It's the only dependency that is statically linked into the core LWJGL shared library.

The only workaround I can think of is going back to libffi, but I'm not looking forward to it at all (it's a pita to build on Windows).

**Comment by brunoborges** [comment]
Let me bring this up to our engineering team and see if we can help on this `dyncall` port.

Thank you so much for the excellent and detailed summary of what needs to be done!

// @microsoft/openjdk-aarch

**Comment by lewurm** [comment]
Thank you @Spasi, that is really helpful.

I'll look into the `dyncall` port to Windows+Arm64 probably next week, and report back.

**Comment by fiddlerwoaroof** [comment]
I can test macOS ARM64 with Minecraft, if there's a build available

**Comment by r58Playz** [comment]
It seems like someone got LWJGL to build according to [this](https://gist.github.com/tanmayb123/d55b16c493326945385e815453de411a).

**Comment by lewurm** [comment]
Sorry for the late update.

### dyncall on macOS+AArch64

I can confirm that `dyncall` works as-is on macOS+AArch64 (aka. Apple Silicon). There are a few test bugs, I included fixes for them in the patchset below.

### dyncall on Windows+AArch64

They following things had to be done:
1. workaround lack of ARMASM64 support in CMake
2. port existing AArch64 assembly snippets to ARMASM64 syntax
3. support var args on Windows+AArch64

The patches are here (generated via `hg export`): https://gist.githubusercontent.com/lewurm/63a2c93a13d20c3ca53cfe7049dbfd48/raw/e03e859b57fae929ce6ffed5684aa7424a78296f/aarch64-fun.export.patch

@Spasi I'm not sure how you go about https://github.com/LWJGL-CI/dyncall, but I assume you manually import it? I sent the patches to the dyncall maintainers, let's see if they merge it upstream at https://dyncall.org/pub/dyncall/dyncall/ . Also the CI config for Windows+AArch64 needs a slight update: Due to the CMake workaround, it's required to call `vcvarsall.bat amd64_arm64` before the first `cmake` call. I didn't know where to PR such change.

### (baby steps) Minecraft on Windows+AArch64

Based on the Apple Silicon instructions that have been posted here ( https://gist.github.com/tanmayb123/d55b16c493326945385e815453de411a ), I tried to port it to Windows+AArch64: https://gist.github.com/lewurm/1b028bf3ebe46b46e406abc00bb580b1 .
Changes:
* it's supposed to be executed in WSL
* swap `:` to `;` in class path string
* convert each path from WSL to Windows via `wslpath -w`
* download/copy native `.dll`s for Windows+AArch64 into `./lwjglnatives/`

Alas the `glfw.dll` posted by @Spasi doesn't cut it, it fails to initialize some OpenGL context (crash log in the gist). I tried to look into it, but that's not the abstraction layer I usually operate in; I was quite lost to say the least 🙂 Maybe someone is willing to look into it further, now that `dyncall` is unblocked.

**Comment by r58Playz** [comment]
Looks like you need to install the [OpenGL compatibility pack.](https://devblogs.microsoft.com/directx/announcing-the-opencl-and-opengl-compatibility-pack-for-windows-10-on-arm/)

**Comment by dustContributor** [comment]
Ah cool, OpenGL is literally not supported on Windows 10 for arm64, unless it's exposed as a GL->D3D12 layer, with all the caveats it implies. Well at least it should be enough for Minecraft, but nothing else for now (ie, no GL4, nor GL 3 with commonly supported GL 4 extensions, possibly no performance improvement mods either since some of them use more advanced GL features).

**Comment by r58Playz** [comment]
GL is supported but the version is ancient(GL 1.2) so no shaders.

**Comment by SaadAbbassi4** [comment]
> GL is supported but the version is ancient(GL 1.2) so no shaders.
> 
> EDIT: Optifine works on Apple Silicon.

Could you please tell me how did you make it work?

**Comment by r58Playz** [comment]
Use [this](https://github.com/yusefnapora/m1-multimc-hack) to get MultiMC(the x86_64 version) to run vanilla 1.16.
After that, install Forge on a 1.16 instance. Add Optifine HD_U_G5 to the instance. I added `-Dfml.earlyprogresswindow=false` as a JVM argument, though [it might not work for you](https://github.com/MinecraftForge/MinecraftForge/issues/7558). 
[optionsof.txt](https://github.com/LWJGL/lwjgl3/files/5753086/optionsof.txt)
I attached my Optifine options file.

**Comment by r58Playz** [comment]
By the way, can LWJGL2 get ported to `arm64`?

**Comment by lewurm** [comment]
> Looks like you need to install the [OpenGL compatibility pack.](https://devblogs.microsoft.com/directx/announcing-the-opencl-and-opengl-compatibility-pack-for-windows-10-on-arm/)

Thank you for that @r58Playz! That indeed brings me a bit further with a black window popping up, but the application is shutting down after loading `lwjgl_stb.dll` (not crashing, or least it doesn't look like it): https://gist.github.com/lewurm/c05450a77e6eacab0dc0f70896d1e21b

I run with `-Dorg.lwgl.util.Debug=true -Dorg.lwjgl.util.DebugLoader=true`. Are there other flags that would help me to track down what is going on?

**Comment by r58Playz** [comment]
`jemalloc.dll` is not found. That's the issue.

**Comment by lewurm** [comment]
Thank you for your input @r58Playz!

> `jemalloc.dll` and `opengl32.dll` are not found. Not all the natives for LWJGL are added to the library folder.

Is that really the problem?

>[LWJGL] [14:15:52] [Render thread/INFO]: [STDERR]: Warning: Failed to instantiate memory allocator: org.lwjgl.system.jemalloc.JEmallocAllocator. Using the system default.
> [LWJGL] [14:15:52] [Render thread/INFO]: [STDERR]: MemoryUtil allocator: StdlibAllocator

https://github.com/LWJGL/lwjgl3/blob/3c92f417252da6f2b6cbfba75db9e4c62bc28a9e/modules/lwjgl/core/src/main/java/org/lwjgl/system/MemoryManage.java#L25-L53

That looks okay to me. Am I missing something?

And regarding opengl:
> [14:15:52] [Render thread/INFO]: [STDERR]: 	Loaded from system paths: C:\WINDOWS\SYSTEM32\opengl32.dll

Sounds okay to me as well.

**Comment by r58Playz** [comment]
Oh, the OpenGL part was my mistake but there aren't any other issues in the log.

**Comment by r58Playz** [comment]
Try running without any arguments.

**Comment by Spasi** [comment]
Hello, an update:

The dyncall bindings in LWJGL have now been replaced with libffi. Apologies to @lewurm and anyone else that spent time on getting dyncall to work on ARM64, your effort is much appreciated. However, I believe this was a necessary step towards supporting more platforms/architectures (libffi supports [virtually everything](https://github.com/libffi/libffi/blob/master/README.md#supported-platforms) we could ever need). This migration has also resolved a range of long-standing issues in LWJGL (e.g. dyncall couldn't handle structs returned by-value) and the new support for upcalls is both simpler and more efficient.

For users that would like to help with the migration process or simply want to experiment with LWJGL, the nightly build (https://www.lwjgl.org/browse/nightly) now includes an up-to-date libffi static library for all current+future supported platform/architectures:

- Linux: arm32, arm64, mips64, x64
- macOS: arm64, x64
- Windows: arm64, x64, x86

Next step is updating the CI infrastructure. Unfortunately, there was a major setback recently, with the [changes](https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing) to Travis CI's pricing model. The new plan is to migrate everything to Github Actions. I've made some testing with libffi (workflow [here](https://github.com/LWJGL-CI/libffi/blob/master/.github/workflows/lwjgl.yml)) and I'm happy with what I'm seeing so far. It will be great having everything in one place (vs some on Travis CI, some on AppVeyor).

A couple of issues atm:

- The macOS/ARM64 build on Github Actions worked fine when I first set it up, but is no longer available. The `macos-11.0` runner is in private preview atm and I'm not sure when it's going to be available again.
- I can't get libffi to build on Windows (on any architecture). Basically, I've been trying to port [this AppVeyor script](https://github.com/LWJGL-CI/libffi/blob/a8faf543547de21e0158a8ab6fcdce3a69328237/appveyor.yml), which builds just fine (output [here](https://ci.appveyor.com/project/LWJGL-CI/libffi/builds/37903763/job/vt79v2nsoefiuwd9)), to Github Actions. Any tips on what I may be doing wrong will be appreciated!

**Comment by Spasi** [comment]
The migration is moving along nicely. I found solutions to the above two issues and the following projects are now building with Github Actions:

- [libffi](https://github.com/LWJGL-CI/libffi/actions/runs/605626696)
- [GLFW](https://github.com/LWJGL-CI/glfw/actions/runs/605894942)
- [OpenAL Soft](https://github.com/LWJGL-CI/openal-soft/actions/runs/605895188)
- [Assimp](https://github.com/LWJGL-CI/assimp/actions/runs/605895329)
- [bgfx](https://github.com/LWJGL-CI/bgfx/actions/runs/606120267) (Windows ARM64 currently blocked by bkaradzic/bgfx#2351)

**Comment by Spasi** [comment]
We're [all green](https://github.com/LWJGL-CI/lwjgl3/actions/runs/628122209)! :tada:

Users with access to ARM hardware should now be able to build and test LWJGL 3 locally on macOS and Windows. Please open new issues (or contact me on [Slack](http://slack.lwjgl.org/)/[Discord](https://discord.gg/6CywMCs)) if you face any trouble.

The first `3.3.0` snapshot will also be available very soon (hopefully tomorrow).

**Comment by Spasi** [comment]
LWJGL `3.3.0` snapshot `7` is now available.

The [build customizer](https://www.lwjgl.org/customize) has been updated to include the two new architectures. You can now use it to download a bundle with the natives or get them via Maven/Gradle (the new classifiers are `natives-macos-arm64` and `natives-windows-arm64`).

**Comment by Helic0pter** [comment]
Thanks for your work. Currently I check LWJGL 3.3.0 snapshot 7 with the jmonkeyengine. I have problems with the initialization of OpenAL. In jme3, this works like this:
    public void createALC() {
        device = ALC10.alcOpenDevice((ByteBuffer) null);
        ALCCapabilities deviceCaps = ALC.createCapabilities(device);
        context = ALC10.alcCreateContext(device, (IntBuffer) null);
        ALC10.alcMakeContextCurrent(context);
        AL.createCapabilities(deviceCaps);
    }

I get then:
java.lang.NullPointerException
	at org.lwjgl.system.Checks.check(Checks.java:188)

I guess, you see immediately, how this has to be changed? Thanks!

**Comment by Helic0pter** [comment]
I just saw, that the OpenAL.dll in 3.2.3 is about 1MB, here it is only 300 KB. Could there be a problem i the build process? (Windows 64 bit and also Windows 32 bit)

**Comment by Helic0pter** [comment]
Ok, on Windows it is so that the new lwjgl-openal.jar works with the OLD OpenAL.dll on Windows 10. OLD means from 3.2.3 stable. Fr me it looks as if in the generation of the OpenAL binaries, there is something wrong ...

**Comment by Spasi** [comment]
LWJGL 3.3.0 snapshot 8 is available and the OpenAL-Soft natives will now work on Windows. Thanks for reporting this @Helic0pter!

**Comment by adiantek** [comment]
I ran Minecraft on Surface Pro X. Details: https://github.com/MultiMC/MultiMC5/issues/3025#issuecomment-807908542
I wrote here bcuz it's the first result when I type `minecraft windows aarch64` in Google
